// crownstone-core/dist/util/logging/Log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const enums_1 = require("../../declarations/enums");
class Logger {
    constructor(level) {
        this.levelConfig = enums_1.LogConfig;
        this.level = level;
        this.levelPrefix = this._getPrefix(level);
    }
    setLogLevel(levels) {
        this.levelConfig = levels;
    }
    log(...any) {
        this._log('Info -----', this.levelConfig.log, arguments);
    }
    ble(...any) {
        this._log('BLE ------', this.levelConfig.ble, arguments);
    }
    usb(...any) {
        this._log('USB ------', this.levelConfig.usb, arguments);
    }
    event(...any) {
        this._log('Event ----', this.levelConfig.events, arguments);
    }
    cloud(...any) {
        this._log('Cloud ----', this.levelConfig.cloud, arguments);
    }
    system(...any) {
        this._log('System ---', this.levelConfig.system, arguments);
    }
    error(...any) {
        this._logType('ERROR ----', this.levelConfig.log, enums_1.LOG_LEVEL.error, arguments);
    }
    warn(...any) {
        this._logType('WARNING --', this.levelConfig.log, enums_1.LOG_LEVEL.warning, arguments);
    }
    info(...any) {
        this._logType('Info -----', this.levelConfig.log, enums_1.LOG_LEVEL.info, arguments);
    }
    debug(...any) {
        this._logType('Debug ----', this.levelConfig.log, enums_1.LOG_LEVEL.debug, arguments);
    }
    verbose(...any) {
        this._logType('Verbose --', this.levelConfig.log, enums_1.LOG_LEVEL.debug, arguments);
    }
    _getPrefix(level) {
        switch (level) {
            case enums_1.LOG_LEVEL.verbose:
                return 'v';
            case enums_1.LOG_LEVEL.debug:
                return 'd';
            case enums_1.LOG_LEVEL.info:
                return 'i';
            case enums_1.LOG_LEVEL.warning:
                return 'w';
            case enums_1.LOG_LEVEL.error:
                return 'e';
            default:
                return 'v';
        }
    }
    _log(type, configCheckField, allArguments) {
        if (configCheckField <= this.level) {
            let args = ['LOG' + this.levelPrefix + ' ' + type + ' :'];
            for (let i = 0; i < allArguments.length; i++) {
                let arg = allArguments[i];
                args.push(arg);
            }
            console.log.apply(this, args);
        }
    }
    _logType(type, configCheckField, forcedLevel, allArguments) {
        if (configCheckField <= forcedLevel) {
            let args = ['LOG' + this._getPrefix(forcedLevel) + ' ' + type + ' :'];
            for (let i = 0; i < allArguments.length; i++) {
                let arg = allArguments[i];
                args.push(arg);
            }
            console.log.apply(this, args);
        }
    }
}
exports.Logger = Logger;
exports.LOGv = new Logger(enums_1.LOG_LEVEL.verbose);
exports.LOGd = new Logger(enums_1.LOG_LEVEL.debug);
exports.LOGi = new Logger(enums_1.LOG_LEVEL.info);
exports.LOG = new Logger(enums_1.LOG_LEVEL.info);
exports.LOGw = new Logger(enums_1.LOG_LEVEL.warning);
exports.LOGe = new Logger(enums_1.LOG_LEVEL.error);
